merlin/router.py

```diff
@@ -236,43 +240,130 @@ def create_config(task_server: str, config_dir: str, broker: str, test: str) ->
+    :returns: True if workers are still processing tasks, False otherwise
+    """
+    result = False
+
+    if task_server == "celery":
+        from merlin.celery import app
+
+        result = check_celery_workers_processing(queues_in_spec, app)
+    else:
+        LOG.error("Celery is not specified as the task server!")
+
+    return result
+
+
+def check_merlin_status(args: "Namespace", spec: "MerlinSpec") -> bool:  # noqa
+    """
+    Function to check merlin workers and queues to keep the allocation alive
 
     :param `args`: parsed CLI arguments
-    :param `spec`: the parsed spec.yaml
+    :param `spec`: the parsed spec.yaml as a MerlinSpec object
+    :returns: True if there are still tasks being processed, False otherwise
     """
+    # Initialize the variable to track if there are still active tasks
+    active_tasks = False
+
+    # Get info about jobs and workers in our spec from celery
     queue_status = query_status(args.task_server, spec, args.steps, verbose=False)
+    LOG.debug(f"Monitor: queue_status: {queue_status}")
 
+    # Count the number of jobs that are active
+    # (Adding up the number of consumers in the same way is inaccurate so we won't do that)
     total_jobs = 0
-    total_consumers = 0
-    for _, jobs, consumers in queue_status:
-        total_jobs += jobs
-        total_consumers += consumers
-
-    if total_jobs > 0 and total_consumers == 0:
-        # Determine if any of the workers are on this allocation
-        worker_names = spec.get_worker_names()
-
-        # Loop until workers are detected.
-        count = 0
-        max_count = 10
-        while count < max_count:
-            # This list will include strings comprised of the worker name with the hostname e.g. worker_name@host.
-            worker_status = get_workers(args.task_server)
-            LOG.info(f"Monitor: checking for workers, running workers = {worker_status} ...")
-
-            check = any(any(iwn in iws for iws in worker_status) for iwn in worker_names)
-            if check:
-                break
-
-            count += 1
-            time.sleep(args.sleep)
-
-        if count == max_count:
-            LOG.error("Monitor: no workers available to process the non-empty queue")
-            total_jobs = 0
-
-    return total_jobs
+    for queue_info in queue_status.values():
+        total_jobs += queue_info["jobs"]
+
+    # Get the queues defined in the spec
+    queues_in_spec = spec.get_queue_list(["all"])
+    LOG.debug(f"Monitor: queues_in_spec: {queues_in_spec}")
+
+    # Get the active queues and the workers that are watching them
+    active_queues = get_active_queues(args.task_server)
+    LOG.debug(f"Monitor: active_queues: {active_queues}")
+
+    # Count the number of workers that are active
+    consumers = set()
+    for active_queue, workers_on_queue in active_queues.items():
+        if active_queue in queues_in_spec:
+            consumers |= set(workers_on_queue)
```

Comment:
Can you explain the bit wise OR operation here?

Reply:
The `|` operator in python can be used for taking the union of two sets. Here, I'm doing an in-place union of the `consumers` set and the set of `workers_on_queue`. For example, if I have:
```
consumers = {"worker1", "worker2", "worker3"}
workers_on_queue = {"worker3", "worker5"}
```
Then, doing the in-place union with `consumers |= workers_on_queue` would change `consumers` value to `{"worker1", "worker2", "worker3", "worker5"}`

Reply:
Gotcha! I hadn’t seen that use case for set union